crates/rustic_core/src/commands/restore.rs

```diff
@@ -0,0 +1,620 @@
+//! `restore` subcommand
+
+use log::{debug, error, info, trace, warn};
+
+use std::{
+    cmp::Ordering,
+    collections::BTreeMap,
+    io::Read,
+    num::NonZeroU32,
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+
+use chrono::{DateTime, Local, Utc};
+use ignore::{DirEntry, WalkBuilder};
+use itertools::Itertools;
+use rayon::ThreadPoolBuilder;
+
+use crate::{
+    error::CommandErrorKind, hash, repository::Indexed, DecryptReadBackend, FileType, Id,
+    IndexedBackend, LocalDestination, Node, NodeStreamer, NodeType, Open, Progress, ProgressBars,
+    ReadBackend, Repository, RusticResult, TreeStreamerOptions,
+};
+
+pub(crate) mod constants {
+    pub(crate) const MAX_READER_THREADS_NUM: usize = 20;
+}
+
+/// `restore` subcommand
+#[allow(clippy::struct_excessive_bools)]
+#[derive(clap::Parser, Debug)]
+pub struct RestoreOpts {
+    /// Remove all files/dirs in destination which are not contained in snapshot.
+    /// WARNING: Use with care, maybe first try this with --dry-run?
+    #[clap(long)]
+    pub delete: bool,
+
+    /// Use numeric ids instead of user/group when restoring uid/gui
+    #[clap(long)]
+    pub numeric_id: bool,
+
+    /// Don't restore ownership (user/group)
+    #[clap(long, conflicts_with = "numeric_id")]
+    pub no_ownership: bool,
+
+    /// Always read and verify existing files (don't trust correct modification time and file size)
+    #[clap(long)]
+    pub verify_existing: bool,
+
+    #[clap(flatten)]
+    pub streamer_opts: TreeStreamerOptions,
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct FileDirStats {
+    pub restore: u64,
+    pub unchanged: u64,
+    pub verified: u64,
+    pub modify: u64,
+    pub additional: u64,
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct RestoreStats {
+    pub file: FileDirStats,
+    pub dir: FileDirStats,
+}
+
+impl RestoreOpts {
+    pub(crate) fn restore<P: ProgressBars, S: Indexed>(
+        &self,
+        file_infos: RestoreInfos,
+        repo: &Repository<P, S>,
+        node: &Node,
+        dest: &LocalDestination,
+        dry_run: bool,
+    ) -> RusticResult<()> {
+        if file_infos.restore_size == 0 {
+            info!("all file contents are fine.");
+        } else if dry_run {
+            repo.warm_up(file_infos.to_packs().into_iter())?;
+        } else {
+            repo.warm_up_wait(file_infos.to_packs().into_iter())?;
+            restore_contents(repo, dest, file_infos)?;
+        }
+
+        if !dry_run {
+            let p = repo.pb.progress_spinner("setting metadata...");
+            self.restore_metadata(repo, dest, node)?;
+            p.finish();
+        }
+
+        Ok(())
+    }
+
+    /// collect restore information, scan existing files and remove superfluous files
+    pub(crate) fn collect_and_remove<P: ProgressBars, S: Indexed>(
```

Comment:
Yes, agree that the restore command is pretty complex on one hand and pretty optimized on the other hand...
I also thought about improving it, but then decided to do minimal changes in order to present the command in the API.

Can we work on the restore command in future?
Yes, it is very ugly that there is no possible repo method which would just analyze what needs to be done. But for instance regarding the removal of existing files, you have two alternatives:
- save the files to-remove in some kind of list (which can be very memory consuming if you delete a lot by this command)
- scan the destination (a the repository tree) a second time which can be quite costly.

I think refactoring this would need a bit more thinking and hence I don't want to do this parallel to the current task.